pkgs/top-level/haskell-packages.nix

```diff
@@ -42,7 +44,10 @@ let
       enableExecutableProfiling = false;
       enableLibraryProfiling = false;
       enableSharedExecutables = false;
-      enableSharedLibraries = false;
+      # Usually necessary to build e.g. QuickCheck (required by hadrian).
+      # This binary GHC built on Alpine is statically linked and can't produce
+      # shared libraries.
+      enableSharedLibraries = self.ghc.name != "ghc-binary-musl-9.2.2";
```

Comment:
Good news, we can likely drop this when [!8699](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/8699) is merged and backported to `ghc-9.4`.